### ISSUE TITLE
Turn on compile time simulation of VM operand stack by default 

### ISSUE THREAD

**Issue by srikanth-sankaran** [eclipse-jdt/eclipse.jdt.core#2234]
Compiling switch expressions that embed a try block comes with some very interesting challenges. In general, when an exception is raised, the VM arranges for the catch block (internally the finally block is just another form of catch block) to be entered with NOTHING but the thrown exception on the operand stack. Any partially cooked expressions and sub-expressions of larger expressions - that were under evaluation at the time of raising of the exception - are all discarded. 

This is fine for a switch statement - when runtime evaluation of a switch statement begins the operand stack is empty (unless the switch statement is embedded in a switch expression but the problem really decays to one of switch expressions being special) - but this is not fine for a switch expression. When evaluation of a switch expression begins, operand stack may contain any number of partially cooked expressions and sub-expressions, that would be needed after the evaluation of the switch expression completes to combine with the yielded value to compute eventual result.

Think of `foo(a, b, c, switch (x) { /* try - catch somewhere here */ });`

So discarding operand stack values on an exception being raised would interfere with the evaluation of the larger expression that embeds the switch expression.

As a result,  compilers resort to spilling the stack to locals and restoring later at chosen points - this requires the operand stack contents to be tracked.

Presently, ECJ tracks the operand stack for a method **IFF** it is composed with a switch expression that contains a try block and not otherwise.

That is a fantastic cover for bugs to go hide under!

See:

https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2228
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2231
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2233
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2318
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2319
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2322
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2323
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2335
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2349
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2363
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2366
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2373
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2387
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2423

These are defects in the track stacking code that went uncaught/undiscovered so far because there is no test case that exercises track stacking code in these scenarios.

I propose that we turn on stack tracking by default - that will heavily exercise the code and expose defects ahead of time.

**Comment by srikanth-sankaran** [comment]
Added benefits:

1. This could be readily enhanced to compare stack contents across branch arc and where they differ report an error i.e., catch verify errors at compile time itself without having to run the program.
2. We may want to explore whether the separate pass that emits stack map table could be hoisted into regular code generation itself

**Comment by iloveeclipse** [comment]
There are 18 new test fails **with target JLS < 1.6** on master after this change, see https://download.eclipse.org/eclipse/downloads/drops4/I20240507-2220/testresults/html/org.eclipse.jdt.core.tests.compiler_ep432I-unit-cen64-gtk3-java17_linux.gtk.x86_64_17.html

- SanityTestEvaluationContext
- NegativeCodeSnippetTest
- VariableTest

Different stacks, one is for example NPE in compiler code, which means it is not test only issue:

```
Cannot read field "declaringClass" because "this.methodDeclaration.binding" is null

java.lang.NullPointerException: Cannot read field "declaringClass" because "this.methodDeclaration.binding" is null
at org.eclipse.jdt.internal.compiler.codegen.CodeStream.retrieveLocalType(CodeStream.java:7664)
at org.eclipse.jdt.internal.compiler.codegen.CodeStream.pushTypeBinding(CodeStream.java:7705)
at org.eclipse.jdt.internal.compiler.codegen.CodeStream.aload_0(CodeStream.java:346)
at org.eclipse.jdt.internal.compiler.codegen.CodeStream.load(CodeStream.java:6275)
at org.eclipse.jdt.internal.compiler.codegen.CodeStream.load(CodeStream.java:6186)
at org.eclipse.jdt.internal.compiler.ast.SingleNameReference.generateCode(SingleNameReference.java:520)
at org.eclipse.jdt.internal.compiler.ast.MessageSend.generateCode(MessageSend.java:618)
at org.eclipse.jdt.internal.compiler.ast.Expression.generateCode(Expression.java:763)
at org.eclipse.jdt.internal.compiler.ast.Block.generateCode(Block.java:87)
at org.eclipse.jdt.internal.compiler.ast.TryStatement.generateCode(TryStatement.java:750)
at org.eclipse.jdt.internal.compiler.ast.Block.generateCode(Block.java:87)
at org.eclipse.jdt.internal.compiler.ast.Initializer.generateCode(Initializer.java:70)
at org.eclipse.jdt.internal.compiler.ast.Clinit.generateCode(Clinit.java:321)
at org.eclipse.jdt.internal.compiler.ast.Clinit.generateCode(Clinit.java:132)
at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.generateCode(TypeDeclaration.java:762)
at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.generateCode(TypeDeclaration.java:832)
at org.eclipse.jdt.internal.compiler.ast.CompilationUnitDeclaration.generateCode(CompilationUnitDeclaration.java:412)
at org.eclipse.jdt.internal.compiler.Compiler.process(Compiler.java:915)
at org.eclipse.jdt.internal.compiler.Compiler.processCompiledUnits(Compiler.java:574)
at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:474)
at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:425)
at org.eclipse.jdt.internal.eval.Evaluator.getClasses(Evaluator.java:132)
at org.eclipse.jdt.internal.eval.EvaluationContext.evaluateVariables(EvaluationContext.java:409)
at org.eclipse.jdt.core.tests.eval.SanityTestEvaluationContext.testEvaluateVariables(SanityTestEvaluationContext.java:168)
```

**Comment by srikanth-sankaran** [comment]
> There are 18 new test fails **with target JLS < 1.6** on master after this change, see https://download.eclipse.org/eclipse/downloads/drops4/I20240507-2220/testresults/html/org.eclipse.jdt.core.tests.compiler_ep432I-unit-cen64-gtk3-java17_linux.gtk.x86_64_17.html
> 
> * SanityTestEvaluationContext
> * NegativeCodeSnippetTest
> * VariableTest
> 
> Different stacks, one is for example NPE in compiler code, which means it is not test only issue:

Thanks for raising this - Looking into it... Reminder to self: for substantial changes don't just rely on github PR testing since that runs at select levels.

**Comment by srikanth-sankaran** [comment]
The jdt-dev discussion thread about withdrawing support for ancient language levels was started because of the plan to simulate operand stack by default only for 1.6+ (See https://www.eclipse.org/lists/jdt-dev/msg02357.html) - but that stalled waiting for feedback from IBM teams on the acceptability of the proposal. 

@mpalat - Can you summarize the position on this ? Can we say only 1.6+ ?
